CMakeLists.txt

```diff
@@ -77,6 +78,30 @@ option(
   OFF)
 option(VELOX_MONO_LIBRARY "Build single unified library." OFF)
 option(ENABLE_ALL_WARNINGS "Enable -Wall and -Wextra compiler warnings." ON)
+option(VELOX_BUILD_SHARED "Build Velox as shared libraries." OFF)
+# While it's possible to build both in one go we currently want to build either
+# static or shared.
+cmake_dependent_option(
+  VELOX_BUILD_STATIC
+  "Build Velox as static libraries."
+  ON
+  "NOT VELOX_BUILD_SHARED"
+  OFF)
+
+if(VELOX_BUILD_SHARED AND NOT VELOX_MONO_LIBRARY)
+  # The large number of targets currently in use within Velox make a shared
+  # build when not using the mono library brittle and labor intensive
+  message(
+    FATAL_ERROR
+      "Shared build is only possible when `VELOX_MONO_LIBRARY=ON` is set.")
+endif()
+
+if(VELOX_BUILD_SHARED)
+  message(
+    WARNING
+      "When building Velox as a shared library it's recommended to build against a shared build of folly to avoid issues with linking of gflags."
```

Comment:
nit: Maybe let user know what they should set to have folly be a shared library ?